gadget/update.go

```diff
@@ -326,6 +327,102 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
+		// at this point all partitions should be created
+		AssumeCreatablePartitionsCreated: true,
+	}
+	if err := EnsureLayoutCompatibility(laidOutVolume, diskLayout, opts); err != nil {
+		return res, fmt.Errorf("volume %s is not compatible with disk %s: %v", vol.Name, dev, err)
+	}
+
+	// also get a Disk{} interface for this device
+	disk, err := disks.DiskFromDeviceName(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot get disk for device %s: %v", dev, err)
+	}
+
+	mappedStructures := make([]DiskStructureDeviceTraits, 0, len(diskLayout.Structure))
+
+	diskPartitions, err := disk.Partitions()
+	if err != nil {
+		return res, fmt.Errorf("cannot get partitions for disk device %s: %v", dev, err)
+	}
+
+	// make an easy map of start offsets to to partitions for lookup
+	diskPartitionsByOffset := map[uint64]disks.Partition{}
+	for _, p := range diskPartitions {
+		diskPartitionsByOffset[p.StartInBytes] = p
+	}
+
+	// create the traits for each structure looping over the laid out structure
+	// to ensure that extra partitions don't sneak in - we double check things
+	// again below this loop
+	for _, structure := range laidOutVolume.LaidOutStructure {
+		part, ok := diskPartitionsByOffset[uint64(structure.StartOffset)]
+		if !ok {
+			// unexpected error - somehow this structure's start offset is not
+			// present in the OnDiskVolume, which is unexpected because we
+			// validated that the laid out volume structure matches the on disk
+			// volume
+			return res, fmt.Errorf("internal error: inconsistent disk structures from LaidOutVolume and disks.Disk: structure starting at %d missing on disk", *structure.Offset)
+		}
+		ms := DiskStructureDeviceTraits{
+			Size:               quantity.Size(part.SizeInBytes),
+			Offset:             quantity.Offset(part.StartInBytes),
+			PartitionUUID:      part.PartitionUUID,
+			OriginalKernelPath: part.KernelDeviceNode,
+			OriginalDevicePath: part.KernelDevicePath,
+			PartitionType:      part.PartitionType,
+			PartitionLabel:     part.PartitionLabel,  // this will be empty on dos disks
+			FilesystemLabel:    part.FilesystemLabel, // blkid encoded
+			FilesystemUUID:     part.FilesystemUUID,  // blkid encoded
+			FilesystemType:     part.FilesystemType,
+		}
+
+		mappedStructures = append(mappedStructures, ms)
+
+		// delete this partition from the map
+		delete(diskPartitionsByOffset, uint64(structure.StartOffset))
+	}
+
+	// check that we don't have any extra partitions on the disk that are not
+	// accounted for in the layout - again this is unexpected since we confirmed
+	// above that the laid out volume was compatible with the on disk volume
+	if len(diskPartitions) != len(mappedStructures) {
+		leftovers := []string{}
+		for _, part := range diskPartitionsByOffset {
+			leftovers = append(leftovers, part.KernelDeviceNode)
+		}
+		return res, fmt.Errorf("internal error: unexpected additional partitions on disk %s not present in the gadget layout: %v", disk.KernelDeviceNode(), leftovers)
```

Comment:
can this happen on already existing devices out there? is there a use case for extra partition like this?

Reply:
Maybe I didn't make this clear enough, but this is really actually an unexpected internal error, this check today is mainly for paranoia and TBH I may have actually written this snippet of code before I had expanded the logic in EnsureLayoutCompatibility to check the same number of partitions on disk with what's in the gadget and vice-versa where this could have happened.

But to be super clear for this to happen today, we need some constellation of possible unlikely bugs right now:
1) EnsureLayoutCompatibility must have a critical bug which allows there the LaidOutVolume and the OnDiskVolume to have different numbers of partitions. EnsureLayoutCompatibility already has checks to ensure that there are not extra on-disk partitions which are not mentioned in the gadget.yaml for a given volume, and also the inverse that there are not extra in-gadget laid out volumes which were not found on disk. We have unit tests for both cases, so if this happens, I would consider this a really really weird and confusing bug inside EnsureLayoutCompatibility.
1) The number of on-disk partitions would need to change between line 339 and the call to OnDiskVolumeFromDevice and line 355 with the call to DiskFromDeviceName. The specific situation would be where on line 339, we have the right/expected set of partitions, and on line 355 somehow we have an extra partition or we lost a partition, such that EnsureLayoutCompatibility was happy, but then re-creating the disks.Disk gives us a different set of partitions on disk that don't match anymore. Looking at this code more closely, actually I think we could eliminate this possibility by having OnDiskVolumeFromDevice also return the disks.Disk (or inversely, having OnDiskVolumeFromDevice accept the disks.Disk instead of finding it itself), since that function actually internally uses DiskFromDeviceName but doesn't return the disks.Disk it creates. That would mean that the only way the on disk partitions could not match what's in the disk is the case of bug 1 above. This wouldn't be a ton of work, but IMHO is marginal gain.


Let me know what you would like me to do about this, but I think this code is fine as-is personally.

Reply:
well hindsight is 20/20 I suppose.

This could actually happen in the case where we make EnsureLayoutCompatibility more permissive as we need to do to allow the implicit system-data role case for UC16/UC18. In this case we will have an extra partition here, and we need to handle it accordingly, so I have added code to do just that with an extra option to this function.